Allow plugins to report their status

mitchelhaan opened the pull request:
This is a proposal to implement #448 with potential for other plugins to provide status as well.

Main changes:
- Each plugin now includes a Status() function
- New config field to enable statistics collection
- Example implementation of status for the cache plugin, reporting cache statistics

Cache statistics:
- entries: number of DNS entries stored in the cache (which could include expired entries)
- hits: queries that were resolved from the cache
- misses: queries that needed to be forwarded
- expirations: queries that were cached but the TTL expired

mitchelhaan committed:
Add config field for plugin status output interval

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Add Status function to all plugins
For now, only the cache plugin has status output

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Add recurring call to plugin Status function

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

jedisct1 commented:
Hi!

And thanks!

This is a good start, but there are a few issues with this approach:

- It introduces some overhead (especially the mutex) to everyone, even when statistics are not enabled.
- Periodically writing unstructured data to a log file may be useful for debugging, not so much for real applications. We can't use this for a web UI, for example. A gRPC interface, chunked HTTP responses, or a standard Prometheus API may be better an option.

mitchelhaan commented:
Yeah, that makes sense. I was mainly concerned with collecting the data and hadn't come up with a good idea for presenting it yet.
Adding a field for enabling/disabling statistics would be easy enough though.

jedisct1 commented:
I'm planning to write a Pi-Hole replacement. A web UI that shows stats, logs, and let you edit the configuration.

So anything that can be updated live and is easy to parse would be preferable.

mitchelhaan committed:
Use atomic counters instead of mutex, only get entry count on demand

This should reduce the statistics overhead to almost nothing

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Use collect_statistics flag instead of collection interval

Planning for a way for statistics to be collected on demand, rather than interval-based

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan commented:
I'm thinking about an API structure like so:

Global proxy config/info:
`/api/v1/proxy/{status,config,reload?}`

Plugin-specific:
`/api/v1/plugin/<name>/{status,config,reload?}`

So a `GET` request for `/api/v1/plugin/cache/status` would return a JSON object containing the cache plugin status (enabled/disabled, statistics, etc)

mitchelhaan committed:
Remove Status()/statistics collection

Preparing for alternate interface for statistics collection

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Add configuration option for management interfaces

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Add basic http server functionality for management

This should be modified to match the other listeners at some point

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Draft status API for cache plugin

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Merge branch 'master' into feature/plugin-status

mitchelhaan committed:
Just use cacheStats locally, count expirations as misses

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>

mitchelhaan committed:
Add detailed stats to show valid/expired cache entries

Signed-off-by: Mitchel Haan <mitchel@mhstud.io>
